File: 2008Practise/PracticeProblems/AlienNumbers.py
import math

def alienNumbers():

    nCases = 0
    l = 0
    f = open ("C:\\gcj\\2008Practise\\PracticeProblems\\A-small-practice.in",'r')
    g = open ("C:\\gcj\\2008Practise\\PracticeProblems\\A-small-practice.out",'w')
    lines = f.readlines()
    nCases = int(lines[0])
    l = l + 1

    for i in range(1,nCases+1):
        g.write("Case #" + str(i) + ": " + translateNumber(lines[l].split()) + "\n")
        l = l+1

    return 1

def translateNumber(lineParts):
    number = list(lineParts[0])
    numLen = len(number)
    sourceLang = list(lineParts[1])
    targetLang = list(lineParts[2])
    newBase = len(targetLang)
    base = len(sourceLang)
    valueToTranslate = 0
    outNumb = ''
    for i in range(0,numLen):
        valueToTranslate = valueToTranslate + base**(numLen-(i+1))*sourceLang.index(number[i])
    largestExponent = math.floor(math.log(valueToTranslate)/math.log(newBase))
    for i in range(int(largestExponent),-1,-1):
        if ((newBase**(i+1) > valueToTranslate) and (newBase**(i) <= valueToTranslate)):
            for j in range(0,newBase):
                if ((j+1)*newBase**i > valueToTranslate):
                     valueToTranslate = valueToTranslate - j*newBase**i
                     outNumb = outNumb + targetLang[j]
                     break
        else:
            outNumb = outNumb + targetLang[0]

    return outNumb

6**3*2+6*3+4
alienNumbers()
6-5
"hej" + str(3)
range(3,1,-1)

File: 2008Practise/PracticeProblems/AlienNumbers.clj
(ns alienNumbers
  (:use [clojure.java.io :only [reader writer]]
        [clojure.string :only [split]])
  (:import (java.io File)))

(Math/pow 2 3)

(defn read-lines [file]
  (with-open [r (reader file)]
    (doall (line-seq r))))

(defn write-lines [file s]
  (with-open [w (writer file)]
    (doseq [line s]
      (.write w line))))


(defn to-base-ten [number source-lang base]
  (let [num-length (count number)]
    (loop [sum 0 i 0]
      (if (= i num-length)
        sum
        (do
          (print sum)
        (recur (+ sum (* (Math/pow base (- num-length (+ base 1))) (.indexOf source-lang (str (nth number i))))) (inc i)) )))))

(defn translate-number [number source-lang target-lang]
  (let [num-len (count number)
        base (count source-lang)
        new-base (count target-lang)
        value-to-translate (to-base-ten number source-lang base)
        largest-exponent (Math/floor (/ (Math/log value-to-translate) (Math/log new-base)))]
    value-to-translate))

(defn alien-numbers []
  (let [lines (read-lines "c:\\gcj\\2008Practise\\PracticeProblems\\A-small-practice.in")
        nCases (read-string (first lines))]
   (doseq [l [(second lines)]]
     (let[l-parse (split l #" ")
          out (translate-number (first l-parse) (second l-parse) (nth l-parse 2))]
       (print out)))))
;       (write-lines "c:\\gcj\\2008Practise\\PracticeProblems\\A-small-practice.out.clj" out)))))

(alien-numbers)
(.indexOf "hej" (str(second "he")))